presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -364,6 +412,243 @@ else if (exchange.getType() == ExchangeNode.Type.REPARTITION) {
+
+            PartitioningHandle partitioningHandle = partitioningScheme.getPartitioning().getHandle();
+            ConnectorId connectorId = partitioningHandle.getConnectorId()
+                    .orElseThrow(() -> new IllegalArgumentException("Unsupported partitioning handle: " + partitioningHandle));
+
+            Map<Symbol, Expression> constantExpressions = new HashMap<>();
+            Partitioning partitioning = partitioningScheme.getPartitioning();
+            List<Symbol> partitioningSymbols = assignPartitioningSymbols(partitioning, constantExpressions);
+            Map<Symbol, ColumnMetadata> temporaryTableColumns = assignColumnNames(exchange.getOutputSymbols(), constantExpressions.keySet());
+            List<String> partitioningColumns = partitioningSymbols.stream()
+                    .map(symbol -> temporaryTableColumns.get(symbol).getName())
+                    .collect(toImmutableList());
+            PartitioningMetadata partitioningMetadata = new PartitioningMetadata(partitioningHandle, partitioningColumns);
+
+            TableHandle temporaryTable = metadata.createTemporaryTable(
+                    session,
+                    connectorId.getCatalogName(),
+                    ImmutableList.copyOf(temporaryTableColumns.values()),
+                    Optional.of(partitioningMetadata));
+
+            TableScanNode scan = createTemporaryTableScan(
+                    temporaryTable,
+                    exchange.getOutputSymbols(),
+                    temporaryTableColumns,
+                    partitioningMetadata);
+
+            TableFinishNode write = createTemporaryTableWrite(
+                    temporaryTable,
+                    temporaryTableColumns,
+                    exchange.getOutputSymbols(),
+                    exchange.getInputs(),
+                    exchange.getSources(),
+                    constantExpressions,
+                    partitioningMetadata,
+                    exchange.getPartitioningScheme().isReplicateNullsAndAny());
+
+            FragmentProperties writeProperties = new FragmentProperties(new PartitioningScheme(
+                    Partitioning.create(SINGLE_DISTRIBUTION, ImmutableList.of()),
+                    write.getOutputSymbols()));
+            writeProperties.setCoordinatorOnlyDistribution();
+
+            List<SubPlan> children = ImmutableList.of(buildSubPlan(write, writeProperties, context));
+            context.get().addChildren(children);
+
+            return scan;
+        }
+
+        private List<Symbol> assignPartitioningSymbols(Partitioning partitioning, Map<Symbol, Expression> constantPartitioningSymbols)
+        {
+            ImmutableList.Builder<Symbol> result = ImmutableList.builder();
+            for (ArgumentBinding argumentBinding : partitioning.getArguments()) {
+                Symbol symbol;
+                if (argumentBinding.isConstant()) {
```

Comment:
hmm... what about these constant symbols used for?...

Reply:
Discussed offline